File: 6_content_filters/c++98/ingredient_application.cxx
/*
 * (c) Copyright, Real-Time Innovations, 2020.  All rights reserved.
 * RTI grants Licensee a license to use, modify, compile, and create derivative
 * works of the software solely for use with RTI Connext DDS. Licensee may
 * redistribute copies of the software provided that all such copies are subject
 * to this license. The software is provided "as is", with no warranty of any
 * type, including any warranty for fitness for any purpose. RTI is under no
 * obligation to maintain or support the software. RTI shall not be liable for
 * any incidental or consequential damages arising out of the use or inability
 * to use the software.
 */

#include <iostream>
#include <stdio.h>
#include <stdlib.h>

#include "chocolate_factory.h"
#include "chocolate_factorySupport.h"
#include "ndds/ndds_cpp.h"
#include "application.h"

using namespace application;

// Ingredient application:
// 1) Subscribes to the lot state
// 2) "Processes" the lot. (In this example, that means sleep for a time)
// 3) After "processing" the lot, publishes the lot state


static int shutdown(
        DDSDomainParticipant *participant,
        const char *shutdown_message,
        int status);

void process_lot(
        const StationKind station_kind,
        ChocolateLotStateDataReader *lot_state_reader,
        ChocolateLotStateDataWriter *lot_state_writer)
{
    // Take all samples.  Samples are loaned to application, loan is
    // returned when LoanedSamples destructor called.
    ChocolateLotStateSeq data_seq;
    DDS_SampleInfoSeq info_seq;

    // Take available data from DataReader's queue
    DDS_ReturnCode_t retcode = lot_state_reader->take(data_seq, info_seq);
    
    if (retcode != DDS_RETCODE_OK && retcode != DDS_RETCODE_NO_DATA) {
        std::cerr << "take error " << retcode << std::endl;
        return;
    }

    // Process lots waiting for ingredients
    for (int i = 0; i < data_seq.length(); ++i) {
        // Check if a sample is an instance lifecycle event
        if (info_seq[i].valid_data) {
            // No need to check that this is the next station: content filter
            // ensures that the reader only receives lots with
            // next_station == this station.
            std::cout << "Processing lot #" << data_seq[i].lot_id << std::endl;

            // Send an update that the this station is processing lot
            ChocolateLotState updated_state(data_seq[i]);
            updated_state.lot_status = PROCESSING;
            updated_state.next_station = INVALID_CONTROLLER;
            updated_state.station = station_kind;
            DDS_ReturnCode_t retcode =
                    lot_state_writer->write(updated_state, DDS_HANDLE_NIL);
            if (retcode != DDS_RETCODE_OK) {
                std::cerr << "write error " << retcode << std::endl;
            }

            // "Processing" the lot.
            DDS_Duration_t processing_time = { 5, 0 };
            NDDSUtility::sleep(processing_time);

            // Send an update that this station is done processing the lot
            updated_state.lot_status = COMPLETED;
            updated_state.next_station = (StationKind)(station_kind + 1);
            updated_state.station = station_kind;
            retcode = lot_state_writer->write(updated_state, DDS_HANDLE_NIL);
            if (retcode != DDS_RETCODE_OK) {
                std::cerr << "write error " << retcode << std::endl;
            }

        } else {
            // Received instance state notification
            continue;
        }
    }

    // Data sequence was loaned from middleware for performance.
    // Return loan when application is finished with data.
    retcode = lot_state_reader->return_loan(data_seq, info_seq);
    if (retcode != DDS_RETCODE_OK) {
        std::cerr << "return_loan error " << retcode << std::endl;
    }
}

void on_requested_incompatible_qos(DDSDataReader *reader)
{
    DDS_RequestedIncompatibleQosStatus status;
    reader->get_requested_incompatible_qos_status(status);

    std::cout << "Discovered DataWriter with incompatible policy: ";
    if (status.last_policy_id == DDS_RELIABILITY_QOS_POLICY_ID) {
        std::cout << "Reliability" << std::endl;
    }
    if (status.last_policy_id == DDS_DURABILITY_QOS_POLICY_ID) {
        std::cout << "Durability" << std::endl;
    }    
}

StationKind string_to_stationkind(const std::string& station_kind)
{
    if (station_kind == "SUGAR_CONTROLLER") {
        return SUGAR_CONTROLLER;
    } else if (station_kind == "COCOA_BUTTER_CONTROLLER") {
        return COCOA_BUTTER_CONTROLLER;
    } else if (station_kind == "MILK_CONTROLLER") {
        return MILK_CONTROLLER;
    } else if (station_kind == "VANILLA_CONTROLLER") {
        return VANILLA_CONTROLLER;
    }
    return INVALID_CONTROLLER;
}

int run_example(unsigned int domain_id, const std::string& station_kind)
{
    StationKind current_station = string_to_stationkind(station_kind);
    std::cout << station_kind << " station starting" << std::endl;

    // Load XML QoS from a specific file
    DDSDomainParticipantFactory *factory =
	        DDSDomainParticipantFactory::get_instance();
    DDS_DomainParticipantFactoryQos factoryQos;
    DDS_ReturnCode_t retcode = factory->get_qos(factoryQos);
    if (retcode != DDS_RETCODE_OK) {
        return shutdown(NULL, "get_qos error", EXIT_FAILURE);
    }
    const char *url_profiles[1] = { "qos_profiles.xml" };
    factoryQos.profile.url_profile.from_array(url_profiles, 1);
    factory->set_qos(factoryQos);

    // Connext DDS setup
    // -----------------
    // A DomainParticipant allows an application to begin communicating in
    // a DDS domain. Typically there is one DomainParticipant per application.
    // Uses IngredientApplication QoS profile to set participant name.
    DDSDomainParticipant *participant =
            DDSTheParticipantFactory->create_participant_with_profile(
                    domain_id,
                    "ChocolateFactoryLibrary",
                    "IngredientApplication",
                    NULL /* listener */,
                    DDS_STATUS_MASK_NONE);
    if (participant == NULL) {
        return shutdown(participant, "create_participant error", EXIT_FAILURE);
    }

    // Create Topics

    // Register the datatype to use when creating the Topic
    const char *lot_state_type_name =
            ChocolateLotStateTypeSupport::get_type_name();
    retcode = ChocolateLotStateTypeSupport::register_type(
            participant,
            lot_state_type_name);
    if (retcode != DDS_RETCODE_OK) {
        return shutdown(participant, "register_type error", EXIT_FAILURE);
    }

    // A Topic has a name and a datatype. Create a Topic called
    // "ChocolateLotState" with your registered ChocolateLotState data type
    DDSTopic *lot_state_topic = participant->create_topic(
            CHOCOLATE_LOT_STATE_TOPIC,
            lot_state_type_name,
            DDS_TOPIC_QOS_DEFAULT,
            NULL /* listener */,
            DDS_STATUS_MASK_NONE);
    if (lot_state_topic == NULL) {
        return shutdown(participant, "create_topic error", EXIT_FAILURE);
    }
    // Create content-filtered Topic
    DDS_StringSeq parameters(1);
    parameters.length(1);
    char filter_value[256];
    snprintf(filter_value, 256, "\'%s\'", station_kind.c_str());
    std::cout << "about to set filter_value" << std::endl;
    // String sequence owns memory for the strings it contains, must allocate
    parameters[0] = DDS_String_dup(filter_value);
    std::cout << "done" << std::endl;
    DDSContentFilteredTopic *filtered_lot_state_topic =
            participant->create_contentfilteredtopic(
                    "FilteredLotState",
                    lot_state_topic,
                    "next_station = %0",
                    parameters);

    // Create Publisher and DataWriters

    // A Publisher allows an application to create one or more DataWriters
    // Publisher QoS is configured with default QoS
    DDSPublisher *publisher = participant->create_publisher(
            DDS_PUBLISHER_QOS_DEFAULT,
            NULL /* listener */,
            DDS_STATUS_MASK_NONE);
    if (publisher == NULL) {
        return shutdown(participant, "create_publisher error", EXIT_FAILURE);
    }

    // This DataWriter writes data on Topic "ChocolateLotState". DataWriter
    // QoS is configured using ChocolateLotStateProfile QoS profile
    DDSDataWriter *generic_lot_state_writer = 
            publisher->create_datawriter_with_profile(
                    lot_state_topic,
                    "ChocolateFactoryLibrary",
                    "ChocolateLotStateProfile",
                    NULL /* listener */,
                    DDS_STATUS_MASK_NONE);
    if (generic_lot_state_writer == NULL) {
        return shutdown(participant, "create_datawriter error", EXIT_FAILURE);
    }

    // A narrow is a cast from a generic DataWriter to one that is specific
    // to your type. Use the type specific DataWriter to write()
    ChocolateLotStateDataWriter *lot_state_writer =
            ChocolateLotStateDataWriter::narrow(generic_lot_state_writer);
    if (lot_state_writer == NULL) {
        return shutdown(participant, "DataWriter narrow error", EXIT_FAILURE);
    }

    // Create Subscriber and DataReader

    // A Subscriber allows an application to create one or more DataReaders
    // Subscriber QoS is configured with default QoS
    DDSSubscriber *subscriber = participant->create_subscriber(
            DDS_SUBSCRIBER_QOS_DEFAULT,
            NULL /* listener */,
            DDS_STATUS_MASK_NONE);
    if (subscriber == NULL) {
        shutdown(participant, "create_subscriber error", EXIT_FAILURE);
    }

    // This DataReader reads data of type ChocolateLotState on Topic
    // "ChocolateLotState" using ChocolateLotStateProfile QoS profile
    DDSDataReader *generic_lot_state_reader =
            subscriber->create_datareader_with_profile(
                    lot_state_topic,
                    "ChocolateFactoryLibrary",
                    "ChocolateLotStateProfile",
                    NULL,
                    DDS_STATUS_MASK_NONE);
    if (generic_lot_state_reader == NULL) {
        shutdown(participant, "create_datareader error", EXIT_FAILURE);
    }
    // Get status condition: Each entity has a Status Condition, which
    // gets triggered when a status becomes true
    DDSStatusCondition *status_condition =
            generic_lot_state_reader->get_statuscondition();
    if (status_condition == NULL) {
        shutdown(participant, "get_statuscondition error", EXIT_FAILURE);
    }

    // Enable only the statuses we are interested in:
    //   DDS_DATA_AVAILABLE_STATUS, DDS_REQUESTED_INCOMPATIBLE_QOS_STATUS
    retcode = status_condition->set_enabled_statuses(
            DDS_DATA_AVAILABLE_STATUS | DDS_REQUESTED_INCOMPATIBLE_QOS_STATUS);
    if (retcode != DDS_RETCODE_OK) {
        shutdown(participant, "set_enabled_statuses error", EXIT_FAILURE);
    }

    // Create the WaitSet and attach the Status Condition to it. The WaitSet
    // will be woken when the condition is triggered.
    DDSWaitSet waitset;
    retcode = waitset.attach_condition(status_condition);
    if (retcode != DDS_RETCODE_OK) {
        shutdown(participant, "attach_condition error", EXIT_FAILURE);
    }

    // A narrow is a cast from a generic DataReader to one that is specific
    // to your type. Use the type specific DataReader to read data
    ChocolateLotStateDataReader *lot_state_reader =
            ChocolateLotStateDataReader::narrow(generic_lot_state_reader);
    if (lot_state_reader == NULL) {
        shutdown(participant, "DataReader narrow error", EXIT_FAILURE);
    }

    // Main loop, wait for lots
    // ------------------------
    while (!shutdown_requested) {
        // Wait for ChocolateLotState
        std::cout << "Waiting for lot" << std::endl;

        // wait() blocks execution of the thread until one or more attached
        // Conditions become true, or until a user-specified timeout expires.
        DDSConditionSeq active_conditions_seq;
        DDS_Duration_t wait_timeout = { 10, 0 };
        retcode = waitset.wait(active_conditions_seq, wait_timeout);

        // You get a timeout if no conditions were triggered before the timeout
        if (retcode == DDS_RETCODE_TIMEOUT) {
            std::cout << "Wait timed out after 10 seconds." << std::endl;
            continue;
        } else if (retcode != DDS_RETCODE_OK) {
            std::cerr << "wait returned error: " << retcode << std::endl;
            break;
        }

        // Get the status changes to check which status condition
        // triggered the WaitSet to wake
        DDS_StatusMask triggered_mask = lot_state_reader->get_status_changes();

        // If the status is "Data Available"
        if (triggered_mask & DDS_DATA_AVAILABLE_STATUS) {
            process_lot(current_station, lot_state_reader, lot_state_writer);
        }
        if (triggered_mask & DDS_REQUESTED_INCOMPATIBLE_QOS_STATUS) {
            on_requested_incompatible_qos(lot_state_reader);
        }
    }

    // Cleanup
    // -------
    // Delete all entities (DataWriter, Topic, Publisher, DomainParticipant)
    return shutdown(participant, "shutting down", EXIT_SUCCESS);
}

// Delete all entities
int shutdown(
        DDSDomainParticipant *participant,
        const char *shutdown_message,
        int status)
{
    DDS_ReturnCode_t retcode;

    std::cout << shutdown_message << std::endl;

    if (participant != NULL) {
        // This includes everything created by this Participant, including
        // DataWriters, Topics, Publishers. (and Subscribers and DataReaders)
        retcode = participant->delete_contained_entities();
        if (retcode != DDS_RETCODE_OK) {
            std::cerr << "delete_contained_entities error " << retcode
                      << std::endl;
            status = EXIT_FAILURE;
        }

        retcode = DDSTheParticipantFactory->delete_participant(participant);
        if (retcode != DDS_RETCODE_OK) {
            std::cerr << "delete_participant error " << retcode << std::endl;
            status = EXIT_FAILURE;
        }
    }

    return status;
}

int main(int argc, char *argv[])
{
    // Parse arguments and handle control-C
    ApplicationArguments arguments;
    parse_arguments(arguments, argc, argv);
    if (arguments.parse_result == PARSE_RETURN_EXIT) {
        return EXIT_SUCCESS;
    } else if (arguments.parse_result == PARSE_RETURN_FAILURE) {
        return EXIT_FAILURE;
    }
    setup_signal_handlers();

    // Sets Connext verbosity to help debugging
    NDDSConfigLogger::get_instance()->set_verbosity(arguments.verbosity);

    int status = run_example(arguments.domain_id, arguments.station_kind);

    // Releases the memory used by the participant factory.  Optional at
    // application shutdown
    DDS_ReturnCode_t retcode = DDSDomainParticipantFactory::finalize_instance();
    if (retcode != DDS_RETCODE_OK) {
        std::cerr << "finalize_instance error " << retcode << std::endl;
        status = EXIT_FAILURE;
    }

    return status;
}

File: 6_content_filters/csharp/IngredientApplication/IngredientApplication.cs
/*
 * (c) Copyright, Real-Time Innovations, 2020.  All rights reserved.
 * RTI grants Licensee a license to use, modify, compile, and create derivative
 * works of the software solely for use with RTI Connext DDS. Licensee may
 * redistribute copies of the software provided that all such copies are subject
 * to this license. The software is provided "as is", with no warranty of any
 * type, including any warranty for fitness for any purpose. RTI is under no
 * obligation to maintain or support the software. RTI shall not be liable for
 * any incidental or consequential damages arising out of the use or inability
 * to use the software.
 */

using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Omg.Dds.Core;
using Rti.Dds.Core;
using Rti.Dds.Core.Status;
using Rti.Dds.Domain;
using Rti.Dds.Publication;
using Rti.Dds.Subscription;
using Rti.Dds.Topics;

namespace ContentFilter
{
    /// <summary>
    /// Example subscriber application
    /// </summary>
    public class IngredientApplication
    {
        private bool shutdownRequested;

        private readonly Dictionary<StationKind, StationKind> nextStation =
            new Dictionary<StationKind, StationKind>()
            {
                { StationKind.COCOA_BUTTER_CONTROLLER, StationKind.SUGAR_CONTROLLER },
                { StationKind.SUGAR_CONTROLLER, StationKind.MILK_CONTROLLER },
                { StationKind.MILK_CONTROLLER, StationKind.VANILLA_CONTROLLER },
                { StationKind.VANILLA_CONTROLLER, StationKind.TEMPERING_CONTROLLER }
            };

        private StationKind GetNextStation(StationKind currentStation)
            => nextStation[currentStation];

        private void ProcessLot(
            StationKind currentStation,
            DataReader<ChocolateLotState> lotStateReader,
            DataWriter<ChocolateLotState> lotStateWriter)
        {
            using var samples = lotStateReader.Take();
            foreach (var sample in samples.ValidData())
            {
                // No need to check that this is the next station: content filter
                // ensures that the reader only receives lots with
                // next_station == this station
                Console.WriteLine($"Processing lot #{sample.lot_id}");

                // Send an update that the tempering station is processing lot
                var updatedState = new ChocolateLotState(sample)
                {
                    lot_status = LotStatusKind.PROCESSING,
                    next_station = StationKind.INVALID_CONTROLLER,
                    station = currentStation
                };
                lotStateWriter.Write(updatedState);

                // "Processing" the lot.
                Thread.Sleep(5000);

                // Send an update that this station is done processing lot
                updatedState.lot_status = LotStatusKind.COMPLETED;
                updatedState.station = currentStation;
                updatedState.next_station = GetNextStation(currentStation);
                lotStateWriter.Write(updatedState);
            }
        }

        private void RunExample(int domainId, string stationKind)
        {
            if (!Enum.TryParse<StationKind>(stationKind, out var currentStation))
            {
                throw new ArgumentException("Invalid station");
            }

            // A DomainParticipant allows an application to begin communicating in
            // a DDS domain. Typically there is one DomainParticipant per application.
            // Uses TemperingApplication QoS profile to set participant name.
            var qosProvider = new QosProvider("../qos_profiles.xml");

            // By specifying a default library, we can later refer to the
            // profiles without the library name
            qosProvider.DefaultLibrary = "ChocolateFactoryLibrary";

            DomainParticipant participant = DomainParticipantFactory.Instance
                .CreateParticipant(
                    domainId,
                    qosProvider.GetDomainParticipantQos("IngredientApplication"));

            Topic<Temperature> temperatureTopic =
                participant.CreateTopic<Temperature>("ChocolateTemperature");
            Topic<ChocolateLotState> lotStateTopic =
                participant.CreateTopic<ChocolateLotState>("ChocolateLotState");

            ContentFilteredTopic<ChocolateLotState> filteredLotStateTopic =
                participant.CreateContentFilteredTopic(
                    name: "FilteredLot",
                    relatedTopic: lotStateTopic,
                    filter: new Filter(
                        expression: "next_station = %0",
                        parameters: new string[] { $"'{stationKind}'" }));

            Publisher publisher = participant.CreatePublisher();

            // Create DataWriter of Topic "ChocolateLotState"
            // using ChocolateLotStateProfile QoS profile for State Data
            DataWriter<ChocolateLotState> lotStateWriter = publisher.CreateDataWriter(
                lotStateTopic,
                qosProvider.GetDataWriterQos("ChocolateLotStateProfile"));

            Subscriber subscriber = participant.CreateSubscriber();

            // Create DataReader of Topic "ChocolateLotState", filtered by
            // next_station, and using ChocolateLotStateProfile QoS profile for
            // State Data.
            DataReader<ChocolateLotState> lotStateReader = subscriber.CreateDataReader(
                filteredLotStateTopic,
                qosProvider.GetDataReaderQos("ChocolateLotStateProfile"));

            // Monitor the DataAvailable status
            StatusCondition statusCondition = lotStateReader.StatusCondition;
            statusCondition.EnabledStatuses = StatusMask.DataAvailable;
            statusCondition.Triggered +=
                _ => ProcessLot(currentStation, lotStateReader, lotStateWriter);
            var waitset = new WaitSet();
            waitset.AttachCondition(statusCondition);

            while (!shutdownRequested)
            {
                // Wait for ChocolateLotState
                Console.WriteLine("Waiting for lot");
                waitset.Dispatch(Duration.FromSeconds(10));
            }
        }

        /// <summary>
        /// Main function, receiving structured command-line arguments
        /// via the System.Console.DragonFruit package.
        /// For example: dotnet run -- --station-kind SUGAR_CONTROLLER
        /// </summary>
        /// <param name="domainId">The domain ID to create the DomainParticipant</param>
        /// <param name="stationKind">The value of StationKind to publish</param>
        public static void Main(
            int domainId = 0,
            string stationKind = "COCOA_BUTTER_CONTROLLER")
        {
            var example = new IngredientApplication();

            // Setup signal handler
            Console.CancelKeyPress += (_, eventArgs) =>
            {
                Console.WriteLine("Shuting down...");
                eventArgs.Cancel = true; // let the application shutdown gracefully
                example.shutdownRequested = true;
            };

            try
            {
                example.RunExample(domainId, stationKind);
            }
            catch (Exception ex)
            {
                Console.WriteLine("RunExample exception: " + ex.Message);
                Console.WriteLine(ex.StackTrace);
            }
        }
    }

    // TODO: codegen
    public enum StationKind
    {
        INVALID_CONTROLLER,
        COCOA_BUTTER_CONTROLLER,
        SUGAR_CONTROLLER,
        MILK_CONTROLLER,
        VANILLA_CONTROLLER,
        TEMPERING_CONTROLLER
    }

    public enum LotStatusKind
    {
        WAITING,
        PROCESSING,
        COMPLETED
    }

    public class Temperature : IEquatable<Temperature>
    {
        public Temperature()
        {
        }

        public Temperature(int degreesParam)
        {
            degrees = degreesParam;
        }

        public Temperature(Temperature other_)
        {
            degrees = other_.degrees;
        }

        public string sensor_id { get; set; }
        public int degrees { get; set; }

        public bool Equals(Temperature other)
        {
            if (other == null)
            {
                return false;
            }

            if (ReferenceEquals(this, other))
            {
                return true;
            }
            return degrees.Equals(other.degrees);
        }
    }

    public class ChocolateLotState
    {
        public ChocolateLotState()
        {
            station = new StationKind();
            next_station = new StationKind();
            lot_status = new LotStatusKind();
        }

        public ChocolateLotState(uint lot_idParam, StationKind stationParam, StationKind next_stationParam, LotStatusKind lot_statusParam)
        {
            lot_id = lot_idParam;
            station = stationParam;
            next_station = next_stationParam;
            lot_status = lot_statusParam;
        }

        public ChocolateLotState(ChocolateLotState other_)
        {
            lot_id = other_.lot_id;
        }

        public uint lot_id { get; set; }
        public StationKind station { get; set; }
        public StationKind next_station { get; set; }
        public LotStatusKind lot_status { get; set; }
    }
}
